### ISSUE TITLE
Provide API to allow JSON converters/types to describe their associated JSON schema

### ISSUE THREAD

**Issue by captainsafia** [dotnet/runtime#105769]
The current implementation of the `JsonSchemaExporter` generates schemas for types based on its on semantics. There's currently not strategy for types/converters to advertise their associated JSON schemas.

This gap has presented a blocker for a couple of scenarios in ASP.NET Core's OpenAPI support around JSON schema.

- gRPC's JSON transcoding layer needs to represent schemas for types based on the underlying protobuf representation instead of the generated C# implementation.
- Types that implement a custom converter are not able to define a schema and are represented by the catch-all `true` schema.

While both System.Text.Json and ASP.NET Core expose APIs for modifying schemas at different layers of the stack, both approaches are _reactive_ and allow mutating the schema that is generated instead of proactively dictating the schema that should be set.

Solutions that have been discussed in the past includ exposing a new `IJsonSchemaResolver` interface for type owners to implement:

```csharp
public interface IJsonSchemaResolver
{
    JsonSchema GetJsonSchema(JsonTypeInfo typeInfo); 
    JsonSchema GetJsonSchema(JsonPropertyInfo propertyInfo); 
}
```

Or alternatively including a new virtual method on the JsonConverter to define associated schemas:

```csharp
public class JsonConverter
{
    public virtual JsonSchema? GetSchema(JsonSerializerOptions options) => ...
}
```

cc: @JamesNK @eiriktsarpalis

**Comment by stephentoub** [comment]
> This gap has presented a blocker

@captainsafia, as in we need to address this in .NET 9?

**Comment by captainsafia** [comment]
> @captainsafia, as in we need to address this in .NET 9?

Ah, should've been clearer about this.

Not a blocker for .NET 9, but definitely would be good to slot in for .NET 10.

**Comment by gregsdennis** [comment]
> ```c#
> public interface IJsonSchemaResolver
> {
>     JsonSchema GetJsonSchema(JsonTypeInfo typeInfo); 
>     JsonSchema GetJsonSchema(JsonPropertyInfo propertyInfo); 
> }
> ```

There is no `JsonSchema` type shipping with .Net 9, and I'm not aware of one planned for 10.  The current implementation for generation simply generates a `JsonNode`.

**Comment by JamesNK** [comment]
The problem with the proposed API is JSON schemas exist in the context of a document.

For example, schemas in a document reference each other via `$ref`, and schemas are commonly stored together in `$defs` (I don't know if JsonSchemaMapper has a feature to output generated schemas together in `$defs` yet. If you don't, I _guarantee_ that there will be a lot of people asking for this feature. It's the idiomatic way to organize schemas in JSON schema).

These things happen beyond the scope of an individual `JsonSchema` instance. Customizations to schema generation for a `JsonTypeInfo` need to flow out and impact the rest of the document. If `JsonTypeInfo` needs to generate its own subschemas based on types in its properties, people will want:
* Those schemas to be output in `$defs`
* Other places that have the same schema, e.g. a schema for `Person` type, to all reference the common schema in `$defs` with `$ref`.

I think customization must happen before the document is generated and work at the `Type` level. Swashbuckle does this successfully using [`ISerializerDataContractResolver`](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/531c4c00f0fe59534dace984635f71815343bcfb/src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/ISerializerDataContractResolver.cs).

Using Swashbuckle's contract resolver, I can instruct it what the schema output for a type should be at the `Type` level. For example:

The `Any` type has a JsonConverter that customizes its output. I can then make its schema match the custom output by saying its schema is an object, with one known property of `System.String`, and additional properties of type `Google.Protobuf.Value`. The Swashbuckle schema generated then uses this custom contract to generate the desired JSON schema. `Google.Protobuf.Value`. is stored in the idiomatic shared location, and other places in the doc that also use `Google.Protobuf.Value` have a `$ref` to that shared location.

**Comment by gregsdennis** [comment]
The current implementation does not support `$ref`s, despite my arguing for them (see comments in the relevant PRs).  What is currently provided is a stepping stone; just enough to get asp.net over the line.

**Comment by JamesNK** [comment]
Yes, what `JsonSchemaMapper` currently does is very rudimentary in .NET 9.

$ref and $def are needed. As evidence, the large amount of work that ASP.NET Core needs to do to post-process the output of the mapper to make it usable. The mapper should be the one doing the work to place schemas in definitions and reference them. Some post-processing in ASP.NET Core will be required to modify the JSON schema syntax to be compatible with OpenAPI, but it will be much less, and simpler, than what is currently done.

And as soon as you start adding support for these features, you realize that an API to customize schemas individually doesn't work.

**Comment by eiriktsarpalis** [comment]
For context, the reason why such an API wasn't added in .NET 9 is due to the lack of a built-in `JsonSchema` type. Building these abstractions around `JsonNode` is plainly not a good idea in the long run. Once a `JsonSchema` type is available building this feature should be relatively straightforward.

`JsonSchemaExporter` currently does not use `$def` (or `$id`) OOTB in the schemas that it generates. The reason is naming: for a general-purpose naming scheme we would need to use identifiers that either encode the FQN of a given type or generate UUIDs neither of which felt like an acceptable solution. If we do add support for `$def` or `$id` in the future this should involve exposing an abstraction for users to define their own naming scheme.

**Comment by Peter-B-** [comment]
I try to generate a Json schema for some types, which contain simple value objects generated with [Vogen](https://github.com/SteveDunn/Vogen). This makes me run into the issue described here. 

It would be great to have an interface for describing my schema. For the simple use-case it might be sufficient to override `JsonConverter.GetSchema`, even though I definitely see the value of generating and reusing `$ref`.

Another issue related to this is, that `GetJsonSchemaAsNode` returns `true` for any type with a custom `JsonConverter` defined. 

```json
{
  "type": "object",
  "properties": {
    "propertyWithCustomConverter": true
  }
}
```

I think this violates the JsonSchema specification, [which states](https://json-schema.org/draft/2020-12/json-schema-core#section-10.3.2.1)
> The value of "properties" MUST be an object. Each value of this object MUST be a valid JSON Schema.

I'd prefer to receive an `NotSupportedException`, preferably with a flag to ignore that in `JsonSchemaExporterOptions`.

Does it make sense to raise an issue for that? Or is it too late anyway, with .Net 9 just being released?

**Comment by eiriktsarpalis** [comment]
> I think this violates the JsonSchema specification, [which states](https://json-schema.org/draft/2020-12/json-schema-core#section-10.3.2.1)
> 
> > The value of "properties" MUST be an object. Each value of this object MUST be a valid JSON Schema.

The value of "properties" _is_ an object in your example. It's just that the schemas of individual properties can boolean per the spec.

**Comment by Peter-B-** [comment]
Ah. True. Sorry, I didn't get that.

Besides that, I'm not sure if it is a very helpful property description.

**Comment by gregsdennis** [comment]
@Peter-B- if you think the spec needs to be classified, please feel free to open an issue in the JSON Schema spec repo.

**Comment by Peter-B-** [comment]
Hi @gregsdennis,

thank you, I will consider this.

Sorry for bringing this up in the first place, since this was not my intended contribution to this issue. 

I basically want to second this request for JsonConverters to describe their schema, and bring my use case of the (Vogen-generated) value objects to the table. I see the technical problems with designing such an API, but I think it could enable some interesting use cases.

**Comment by niemyjski** [comment]
Any idea when this will be moved forward? Anything the community can do? We are missing a lot of models from OpenAPI generation due to this.

**Comment by gregsdennis** [comment]
If you're open to using JSON Schema.Net (and its generation extension package), I've solved supporting JSON Schema request validation in ASP.net.

https://blog.json-everything.net/posts/json-schema-aspdotnet/

This has all the code you need to make it work. 

I've been using it at my job for a month, now, and it's really nice.  Soon our entire API (internal and external) will be using it.

Generation of the OpenAPI is still something I'm working on, though.

**Comment by akade** [comment]
Just stumbled across this issue as well and would also appreciate are more integrated way. For anybody interested, I could work around it by `TransformSchemaNode`:

```csharp
TransformSchemaNode = static (context, node) =>
{
    if(context.PropertyInfo?.CustomConverter is YourCustomConverter)
    {
        if (node.GetValueKind() != JsonValueKind.Object) // see note below
        {
            node = new JsonObject();
        }
        ...
    }
}
```

The passed `node` will switch from being a `JsonObject` to `true` as soon as you put your custom converter on it (my property was of type `DateOnly?`). Unless you have something else on the property, like `[Description]`. 

This confused me, couldn't that be an empty `JsonObject` instead of `true`? I would have found that much more intuitive and ergonomic.

**Comment by eiriktsarpalis** [comment]
> This confused me, couldn't that be an empty `JsonObject` instead of `true`? I would have found that much more intuitive and ergonomic.

The recommendation from JSON schema spec authors was to favor `true` of empty objects in schema generation. I've found [this helper](https://github.com/dotnet/extensions/blob/da94b6045fef85f78a5eee183579221cee684e51/src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.Create.cs#L372-L390) to be useful in the context of authoring transformers.

**Comment by wertzui** [comment]
Is there anything planned for .Net 10?
I'm running into problems when providing a return JSON schema for an MCP tool which returns a type which has a custom `JsonConverter`.

I could probably wait for .Net 10, but 11 is too far away so I would need to come up with a workaround.

**Comment by wertzui** [comment]
Upon further digging into the source code, I found that `JsonConverter` already exposes a method:

```csharp
internal virtual JsonSchema GetSchema(JsonNumberHandling numberHandling)
{
	return null;
}
```

So there already is the `JsonSchema` type mentioned by @eiriktsarpalis 
Why has this method not been made public?

**Comment by eiriktsarpalis** [comment]
It's because that `JsonSchema` type is an ad-hoc implementation detail that does not attempt to conform to any of the draft specifications.

**Comment by wertzui** [comment]
Since this is marked for the `11.0.0` milestone and I needed a solution now to get my MCP server to give out meaningful schemata, I'm going to share my work around to get it working.
This is for everyone who is stumbling across the same problem, but please note that it is a hacky solution relying on reflection and will probably break with `11.0.0` or earlier.

First we define an extension method  to set the `Converter` and `EffectiveConverter` on `JsonTypeInfo<T>`.
This ensures that our converter is used during serialization and deserialization.
```csharp
public static class JsonTypeInfoExtensions
{
    private static readonly FieldInfo _converterBackingField = typeof(JsonTypeInfo).GetField($"<{nameof(JsonTypeInfo.Converter)}>k__BackingField", BindingFlags.Instance | BindingFlags.NonPublic) ?? throw new InvalidOperationException("Cannot find backing field for JsonTypeInfo.Converter.");

    /// <summary>
    /// Sets a custom converter on a <see cref="JsonTypeInfo"/>.
    /// Use this when you need to replace the default converter with a custom one for advanced serialization scenarios.
    /// This method uses reflection to set both the public Converter property and the internal EffectiveConverter backing field.
    /// </summary>
    /// <typeparam name="T">The type being serialized/deserialized.</typeparam>
    /// <param name="typeInfo">The JSON type info to set the converter on.</param>
    /// <param name="converter">The custom converter to use. This must not be a converter factory.</param>
    public static void SetConverter(this JsonTypeInfo typeInfo, JsonConverter converter)
    {
        var type = typeInfo.GetType();
        var _effectiveConverterBackingField = type.GetField("<EffectiveConverter>k__BackingField", BindingFlags.Instance | BindingFlags.NonPublic);
        if (_effectiveConverterBackingField is not null)
        {
            var converterType = converter.GetType();
            var fieldType = _effectiveConverterBackingField.FieldType;
            if (!converter.GetType().IsAssignableTo(_effectiveConverterBackingField.FieldType))
                throw new InvalidOperationException($"Cannot set the JSON converter, because the generic types don't match. The JsonTypeInfo is {type.GetPrettyName()} which expects a converter of type {fieldType.GetPrettyName()}, but the passed in converter is of type {converterType.GetPrettyName()}");
            _effectiveConverterBackingField?.SetValue(typeInfo, converter);
        }
        _converterBackingField.SetValue(typeInfo, converter);
    }

    private static string GetPrettyName(this Type type)
    {
        if (!type.IsGenericType)
            return type.Name;

        var genericTypeName = type.Name.Substring(0, type.Name.IndexOf('`'));
        var genericArgs = string.Join(", ", type.GetGenericArguments().Select(GetPrettyName));
        return $"{genericTypeName}<{genericArgs}>";
    }
```

Now we implement `IJsonTypeInfoResolver` in our `JsonConverter<T>` or `JsonConverterFactory`.
You could also create your own class for this, but for me it worked the best doing it this way, so I could reuse private methods of the converter.
```csharp
public class MyConverterFactory: JsonConverterFactory, IJsonTypeInfoResolver
{
    public JsonTypeInfo? GetTypeInfo(Type type, JsonSerializerOptions options)
    {
        // Check your type here and return null if it is not what you want.

        var typeInfo = JsonTypeInfo.CreateJsonTypeInfo(type, options);

        // Add all properties you need. It is important to set Get and Set, but it is not important to set it to anything meaningful, because your converter will be used anyway
        var propertyInfo = typeInfo.CreateJsonPropertyInfo(propertyType, propertyName);
        propertyInfo.Get = (obj) => null;
        propertyInfo.Set = (obj, value) => { };
        propertyInfo.IsGetNullable = isGetNullable;
        propertyInfo.IsSetNullable = isSetNullable;
        typeInfo.Properties.Add(jsonPropertyInfo);

        typeInfo.SetConverter(CreateConverter(type, options));

        return typeInfo;
    }
}
```

In the end we add our implementation to the `JsonOptions`
```csharp
var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
options.TypeInfoResolverChain.Add(new MyConverterFactory());
// Add the default resolver so it can pick up anything without a custom resolver
options.TypeInfoResolverChain.Add(new DefaultJsonTypeInfoResolver());
```

Now you can use `JsonSchemaExporter.GetJsonSchemaAsNode(options, type, exporterOptions)` to get the schema and `JsonSerializer.Serialize(myInstance, options);` to get the serialized JSON. The same goes for `JsonSerializer.Deserialize<Type>(json, options)` for deserialization.

**Comment by gregsdennis** [comment]
@wertzui (and anyone else) I'm curious what you're your need for the schema is.  gRPC uses protobuf, not JSON Schema, so I'm confused as to how it relates. I also don't see any validation performed in your converter.  What is the purpose of getting the schema? 

If you just need a JSON Schema that represents a type and want to be able to explicitly define it (or generate it using attributes), and you don't specifically need it in the context of serialization, I suggest JsonSchema.Net and it's extension JsonSchema.Net.Generation, which will do these things for you.

**Comment by wertzui** [comment]
In my case it is to provide the schema for the tools of my [MCP server](https://github.com/wertzui/AutoMCP).
I'm using [ModelContextProtocol.AspNetCore](https://github.com/modelcontextprotocol/csharp-sdk)s `AIFunction` to create the tools and `AIJsonUtilities.CreateJsonSchema` from [Microsoft.Extensions.AI](https://github.com/dotnet/extensions/blob/7ddf87e16c92efd0d6d4b7da4a9ee7d185e8835f/src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.Create.cs#L169) to create the schema.
This in turn calls [JsonSerializerOptions.GetJsonSchemaAsNode(...)](https://github.com/dotnet/extensions/blob/7ddf87e16c92efd0d6d4b7da4a9ee7d185e8835f/src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.Create.cs#L253C16-L253C53) which uses the `JsonTypeInfo` provided by the `IJsonTypeInfoResolver`s defined in the `JsonOptions`.

In my specific case I need it, because I'm returning [HAL](https://stateless.group/hal_specification.html) from my endpoints.
To specify a `Resource` with a state, the C# class looks like this:
```csharp
public class Resource<T>
{
    public T State { get; set; }
    public ICollection<Link> Links { get; set; }
    public ICollection<Resource> Embedded { get; set; }
}
```
The JSON representation needs to "flatten" the `State` into the resulting object.
Because `System.Text.Json` does not (yet) provide a way to flatten properties into objects (see https://github.com/dotnet/runtime/issues/55120), I'm using a custom `JsonSerializer` to provide this functionality.
And because `JsonSerializerOptions.GetJsonSchemaAsNode(...)` will just return `true` as schema if a custom `JsonSerializer` is defined for a type, I need to hook into the schema generation to provide the correct schema.

**Comment by Peter-B-** [comment]
I use it as a convenient way to generate Json schemas for documentation of inter process communication. I use [Vogen](https://github.com/SteveDunn/Vogen) to get typed objects for IDs, instead of `string`, `int`, or `Guid`, but the Json will contain the underlying type. So for a `CustomerId` property, which might be a  `Guid`, I want to have a Guid string in the Json and describe this in the Schema.

I use the `JsonSchemaExporter`, because it is simple to use and it reflects my `JsonSerializerOptions`.

I'm using the `TransformSchemaNode` to this end, which works fine for now. However, I still think it would be great if a custom `JsonConverter` could return the schema, since it feels like this should be the one to know the details. But I also see the issues of designing the API, as was pointed out above.

**Comment by cleftheris** [comment]
I am in favor of the suggestion having JsonConverters explicitly declare that they affect schema generation. Maybe even having their default behavior not be affected unless otherwise declared and implemented especially for primitive types and strings. I think this issue is easy to miss when creating small projects and can have non deterministic outcome when working with open API until you deep dive inside the System.Text.Json internals. 


Some of the transformers we already made for dotnet 9 and we are porting in dotnet 10 can be seen here If you are interested to see the impact you can our implementation here https://github.com/indice-co/Indice.Platform/tree/develop/src/Indice.AspNetCore/OpenApi/Transformers.

For anyone that is interested in the reasoning and our use cases here is a more in depth explanation:

Additional to what others have pointed out a major issue with the current behavior is that there are many cases for public APIs that need to make the deserialization behavior more forgiving for primitive types and strings. For example the string literal `"true"` or `"false"` is something that we always "polyfill" using a [JsonStringBooleanConverter ](https://github.com/indice-co/Indice.Platform/blob/develop/src/Indice.Common/Serialization/JsonStringBooleanConverter.cs) as we used to do with number literals up until `AllowReadingFromString` was introduced in JsonSerializerOptions. There are numerous other examples for business specific needs that are tackled with custom converters such as for example the number `2026.0` being able to be parsed into an integer etc. All these converters my result in a useless OpenAPI schema since they may interfere even with all primitive properties things such as Enums and even arrays of primitives. 

In our case in order to support OpenAPI we had to recreate many of the out of the box behavior via numerous schema transformers what was a tedious and error prone task.

As a side note I wanted to point out the irony that the framework has gone to great lengths to render a union type for numbers in the openapi generated json schema when the setting mentioned above is set to true `JsonNumberHandling.AllowReadingFromString`. Which is great don't get me wrong but not so critical when using client generation and open API.

**Comment by avin3sh** [comment]
My usecase is similar to https://github.com/dotnet/aspnetcore/issues/63051 / some bits discussed here and would love to see this get shipped in .NET 11. 

so +1 specifically for the Type-level approach JamesNK describes, instead of a per-JsonSchema-instance API. For devs familiar with Swashbuckle's `ISerializerDataContractResolverr` would feel comfortable with this being at the Type level with the generator ultimately emitting one shared component with $refs automatically.

For example, if I had a type implementing `ICollection<T>` and I wanted to represent it as a its own re-usable component (as opposed to having a simple JSON array), in Swashbuckle world I need to go for `MapType` and then additionally  use `IDocumentFilter` (hacky) or wire the whole thing via `ISerializerDataContractResolver` to do it the proper way. In STJ generator world the latter is difficult today without JsonTypeInfo - `JsonSchemaExporter` with `TransformSchemaNode` might fix the shape problem but still doesn't allow a way forward for having a re-usable component. 

Even if the full type-level design can't make it to .NET 11 - even exposing the hook the built-in converters already use would unblock us: currently `JsonConverter.GetSchema(JsonNumberHandling)` is `internal` , so any option for a custom converter is blocked. Making it overridable - returning a `JsonNode` would at least fix the shape for converter-owned types, even if the $defs/$ref componentization comes later.

Whatever the eventual STJ surface, having schema overrides flow at the type level and dealing with $defs/$ref than mutating an already-generated node would be nicer.

**Comment by wertzui** [comment]
I guess the milestone change means, that it won't come for 11.
Maybe for 12...
